go/pkg/ddapi/testutil.go

```diff
@@ -0,0 +1,161 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ddapi
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"text/template"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/testcontainers/testcontainers-go"
+	"github.com/testcontainers/testcontainers-go/wait"
+)
+
+func createJSONBody(metricname string) string {
+	const jsonTextTemplate = `
+	{
+		"series": [
+		  {
+			"metric": "unit.test.metric.{{.metricname}}",
+			"points": [
+			  [
+				{{.secondsSinceEpoch}},
+				1
+			  ]
+			],
+			"tags": [
+			  "version:7.24.1"
+			],
+			"host": "x1carb6",
+			"type": "rate",
+			"interval": 10
+		  }
+		]
+	  }
+`
+	data := map[string]interface{}{
+		"metricname":        metricname,
+		"secondsSinceEpoch": fmt.Sprintf("%v", time.Now().Unix()),
+	}
+
+	// Render template, store result in string.
+
+	t := template.Must(template.New("").Parse(jsonTextTemplate))
+
+	buf := &bytes.Buffer{}
+	if err := t.Execute(buf, data); err != nil {
+		panic(err)
+	}
+
+	// Return buffer (type *bytes.Buffer), can be passed as third argument to
+	// httptest.NewRequest(). This buffer holds UTF-8-encoded JSON. return buf
+
+	// Return string (JSON doc as text). Need to create strings.NewReader(s)
+	// then for passing it as third arg to httptest.NewRequest()
+	return buf.String()
+}
+
+func createTmpDir() (string, error) {
+	// Create a unique temporary directory for this test suite run. Note: "It
+	// is the caller's responsibility to remove the directory when no longer
+	// needed." -- do this in suite teardown.
+	tempdir, err := ioutil.TempDir("/tmp", "opstrace-go-unit-tests")
+	if err != nil {
+		return "", err
+	}
+
+	log.Infof("created suite tmp dir: %s", tempdir)
+	return tempdir, nil
+}
+
+func startCortex(tempdir string) (context.Context, testcontainers.Container, string, error) {
+	ctx := context.Background()
+
+	// Assume that this is the directory that this source file resides in.
+	// That's true when clicking `run test` in vs code, and interestingly also
+	// true when running `make unit-tests`, i.e. when running `go test ...`.
+	// https://stackoverflow.com/a/23847429/145400
+	testdir, err := os.Getwd()
+	log.Infof("test dir: %s", testdir)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Copy the Cortex config file into `tempdir` (under /tmp) so that it can
+	// so that the bindMount source is valid _on the host_, too (assume that
+	// this test runner runs in a container, and that /tmp is shared between
+	// host and containers).
```

Comment:
This docker in docker thing is getting trickier.

Reply:
Yes. I see that there are typos in this comment, will fix

Reply:
change will be here: https://github.com/opstrace/opstrace/pull/290